### PR TITLE
Add support for system text scale factor.

### PR DESCRIPTION
**Pull request by gspencergoog**
This fixes #10524, adding support to the engine and the android platform code for responding to system settings for font scaling factor.

I added these messages to the system channel, as a "systemSettings" type with a single double member called "textScaleFactor".

It triggers the hooks and sets the internal value on the Window object.

### CODE REVIEW

File: shell/common/engine.cc [hunk 1]
```diff
@@ -44,6 +44,7 @@ constexpr char kAssetChannel[] = "flutter/assets";
 constexpr char kLifecycleChannel[] = "flutter/lifecycle";
 constexpr char kNavigationChannel[] = "flutter/navigation";
 constexpr char kLocalizationChannel[] = "flutter/localization";
+constexpr char kSystemChannel[] = "flutter/system";
```
Comment (goderbauer) [inline shell/common/engine.cc]: Since the memory message and the text scale message seem to get processed at two different places it might make sense to have a separate channel for the text scale after all instead of trying to separate the messages on the same system channel?

Do engine folks have an opinion on this? @cbracken @jason-simmons @Hixie

Comment (gspencergoog) [reply shell/common/engine.cc]: I think if I added another channel, I would make it more general than just text scale.  Maybe "flutter/settings" or something.  Is there any place where the different channels are documented (e.g. what they are supposed to be for, what their scope should be, naming conventions, etc)?

Comment (Hixie) [reply shell/common/engine.cc]: I don't think we have such documentation yet, but we should. Please file a bug to make sure we write such docs and add them to the API docs.

Comment (gspencergoog) [reply shell/common/engine.cc]: OK, Done.  Filed [flutter/flutter#12196](https://github.com/flutter/flutter/issues/12196)

Changed the channel to "flutter/settings".

**Comment by gspencergoog** [comment]
So is this good to go?  I want to commit it so I can start rolling the engine to get the flutter side in.

File: lib/ui/testing_hooks.dart [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+/// Do not use for non-testing purposes, or catastrophic problems may (and
+/// probably will) occur.  These APIs are not part of the Flutter API, and are
+/// subject to change.  Do not depend upon their existence.
+class TestingHooks {
+  /// @nodoc
+  /// Make updateTextScaleFactor hook visible for testing.
+  /// Do not use for non-testing purposes: catastrophic problems may occur.
+  static void updateTextScaleFactor(double textScaleFactor) =>
+      _updateTextScaleFactor(textScaleFactor);
+}
```
Comment (Hixie) [inline lib/ui/testing_hooks.dart]: We should definitely not expose an API we don't want people to use.

Comment (Hixie) [reply lib/ui/testing_hooks.dart]: Any API we expose _will_ be used, and as soon as its used we can't break it, regardless of what our documentation says (especially if we hide it with `@nodoc`).

**Comment by Hixie** [comment]
I recommend testing the engine side of this using the part-of hack in <https://github.com/flutter/engine/blob/master/testing/dart/window_hooks_integration_test.dart>. To test whether this actually results in the MediaQuery changing you need probably need a devicelab test where you actually have the system-level font size change or something like that.

**Comment by Hixie** [comment]
cc @jason-simmons for a more detailed review of the C++ side.

File: lib/ui/testing_hooks.dart [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+part of dart.ui;
+
+/// @nodoc
+/// Makes some internal functions visible for internal Flutter testing.
+///
+/// Do not use for non-testing purposes, or catastrophic problems may (and
+/// probably will) occur.  These APIs are not part of the Flutter API, and are
+/// subject to change.  Do not depend upon their existence.
+class TestingHooks {
```
Comment (goderbauer) [inline lib/ui/testing_hooks.dart]: Exposing this seems bad... is there no other way?

File: lib/ui/testing_hooks.dart (hunk 2 above)
Comment (gspencergoog) [reply lib/ui/testing_hooks.dart]: Well, in principle, I agree, but there isn't any other way that I know of to simulate these events.  I discussed this with Jason before implementing it, and he seemed to think this was the best compromise.

Comment (gspencergoog) [reply lib/ui/testing_hooks.dart]: Out of curiosity: why are we _especially_ bound to @nodoc APIs?  Seems like just the opposite to me.

**Comment by gspencergoog** [comment]
Thanks for the integration hooks pointer.  I'll look into using that instead.

File: lib/ui/testing_hooks.dart (hunk 2 above)
Comment (Hixie) [reply lib/ui/testing_hooks.dart]: We talked about this at lunch but for anyone watching:

It's probably not literally true that `@nodoc` is more likely to make people depend on it. There's a combination of factors that could push that way though:
- if we say `@nodoc`, then anything we put in the docs will be missed by anyone who decides to use it.
- people especially like to use secret APIs because it makes them feel cool and hip.
- if we say it's not supported then that lowers our expectation that people will use it, but it doesn't lower the actual probability that people will use it, so the end result is that relative to our expectation, people are more likely to end up depending on it.

At the end of the day though Flutter's convention is to not have "secret" APIs, we either have public documented supported APIs, or we don't expose the API at all, and in particular, our next layer should always use the previous layer's public API so that people can replace any layer and do the same thing we did (this is true at the macro level, e.g. "rendering" vs "dart:ui", as well as the micro level, e.g. Icon using IconTheme).

We want to avoid magic.

**Comment by gspencergoog** [comment]
I signed it!

**Comment by gspencergoog** [comment]
I signed it!

**Comment by gspencergoog** [comment]
OK, well, googlebot aside, I think this is ready, so PTAL.  I've added the integration hooks test, and removed the TestingHooks class.  Hixie and I talked about other ways to test, and decided that it would be really cool to have a method channel that is just for testing (and only enabled in debug builds) that would allow of simulation of engine events.  I'll do that in another PR, though.

File: testing/dart/window_hooks_integration_test.dart [hunk 4]
```diff
@@ -125,65 +140,101 @@ void main() {
...
 
     test('onSemanticsAction preserves callback zone', () {
       Zone innerZone;
       Zone runZone;
+      int action;
 
       runZoned(() {
         innerZone = Zone.current;
-        window.onSemanticsAction = (_, __) {
+        window.onSemanticsAction = (int value, __) {
```
Comment (goderbauer) [inline testing/dart/window_hooks_integration_test.dart]: nit: reduce the number of underscores for second arg?

File: testing/dart/window_hooks_integration_test.dart [hunk 5]
```diff
@@ -55,55 +57,68 @@ void main() {
...
         };
       });
 
       window.onMetricsChanged();
-      _updateWindowMetrics(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
+      _updateWindowMetrics(0.1234, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
       expect(runZone, isNotNull);
       expect(runZone, same(innerZone));
+      expect(devicePixelRatio, isNotNull);
+      expect(devicePixelRatio, equals(0.1234));
```
Comment (goderbauer) [inline testing/dart/window_hooks_integration_test.dart]: Isn't equals(1234) implying that it cannot be null?

Comment (goderbauer) [reply testing/dart/window_hooks_integration_test.dart]: see also a couple of times further down.

File: testing/dart/window_hooks_integration_test.dart [hunk 6]
```diff
@@ -125,65 +140,101 @@ void main() {
...
 
     test('onPlatformMessage preserves callback zone', () {
       Zone innerZone;
       Zone runZone;
+      String name;
 
       runZoned(() {
         innerZone = Zone.current;
-        window.onPlatformMessage = (_, __, ___) {
+        window.onPlatformMessage = (String value, __, ___) {
```
Comment (goderbauer) [inline testing/dart/window_hooks_integration_test.dart]: same

File: testing/dart/window_hooks_integration_test.dart (hunk 5 above)
Comment (gspencergoog) [reply testing/dart/window_hooks_integration_test.dart]: Yeah, I guess you're right.  I add those because when the test fails because it's null, I thought it would be nicer to have it highlight that it was null, but I guess it's pretty obvious for the equals case.

File: testing/dart/window_hooks_integration_test.dart (hunk 4 above)
Comment (gspencergoog) [reply testing/dart/window_hooks_integration_test.dart]: Done.

File: testing/dart/window_hooks_integration_test.dart [hunk 7]
```diff
@@ -125,65 +137,99 @@ void main() {
...
         innerZone = Zone.current;
-        window.onPlatformMessage = (_, __, ___) {
+        window.onPlatformMessage = (String value, _, __) {
           runZone = Zone.current;
+          name = value;
         };
       });
 
-      _dispatchPlatformMessage(null, null, null);
+      _dispatchPlatformMessage("testName", null, null);
```
Comment (Hixie) [inline testing/dart/window_hooks_integration_test.dart]: nit: use single quotes unless there's a reason to use double quotes

File: testing/dart/window_hooks_integration_test.dart [hunk 8]
```diff
@@ -125,65 +137,99 @@ void main() {
...
 
     test('onSemanticsAction preserves callback zone', () {
       Zone innerZone;
       Zone runZone;
+      int action;
 
       runZoned(() {
         innerZone = Zone.current;
-        window.onSemanticsAction = (_, __) {
+        window.onSemanticsAction = (int value, _) {
```
Comment (Hixie) [inline testing/dart/window_hooks_integration_test.dart]: also, not a big deal, but as a general rule we should always include types (even for unused arguments), that way when someone is looking at the code it's more self-documenting, easier to copy/paste, etc. Don't worry about adding the type (and maybe explicit variable name) here unless you happy to be on this line some time, though.

File: testing/dart/window_hooks_integration_test.dart (hunk 7 above)
Comment (gspencergoog) [reply testing/dart/window_hooks_integration_test.dart]: Done.  (It's my years of C++ fingers that need retraining...)

File: testing/dart/window_hooks_integration_test.dart (hunk 8 above)
Comment (gspencergoog) [reply testing/dart/window_hooks_integration_test.dart]: I think for tests that's less of an issue.  I think it adds to readability of the test to be able to just visually skip over unused arguments (because they are so often irrelevant).  I agree that for framework code this is reversed, and that they should always have types.
